data_structures/linked_list/skip_list.py

```diff
@@ -0,0 +1,418 @@
+"""
+Based on "Skip Lists: A Probabilistic Alternative to Balanced Trees" by William Pugh
+https://epaperpress.com/sortsearch/download/skiplist.pdf
+"""
+
+from typing import List, Tuple, Optional, TypeVar, Generic
+from random import random
+
+KT = TypeVar("KT")
+VT = TypeVar("VT")
+
+
+class Node(Generic[KT, VT]):
+    def __init__(self, key: KT, value: VT):
+        self.key = key
+        self.value = value
+        self.forward: List[Node[KT, VT]] = []
+
+    def __repr__(self):
+        """
+        :return: Visual representation of Node
+
+        >>> node = Node("Key", 2)
+        >>> repr(node)
+        'Node(Key: 2)'
+        """
+
+        return f"Node({self.key}: {self.value})"
+
+    @property
+    def level(self) -> int:
+        """
+        :return: Number of forward references
+
+        >>> node = Node("Key", 2)
+        >>> node.level
+        0
+        >>> node.forward.append(Node("Key2", 4))
+        >>> node.level
+        1
+        >>> node.forward.append(Node("Key3", 6))
+        >>> node.level
+        2
+        """
+
+        return len(self.forward)
+
+
+class SkipList(Generic[KT, VT]):
+    def __init__(self, p: float = 0.5, max_level: int = 16):
+        self.head: Node = Node("root", None)
+        self.level: int = 0
+        self.p = p
+        self.max_level = max_level
+
+    def __str__(self):
```

Comment:
A doctest here please and then I will approve.  Nice work!

Reply:
It's hard to write doctest for this method because of randomness but I hope ellipsis solution will fit.